packages/cursorless-vscode/src/keyboard/grammar/getAcceptableTokenTypes.ts

```diff
@@ -43,39 +66,80 @@ function getMetadata<T extends keyof KeyboardCommandHandler>(
     .metadata;
 }
 
+// Prevent infinite recursion
+const MAX_VISITS = 3;
+
+/** Indicates that the given symbol hasn't been typed yet */
+export const MISSING = Symbol("missing");
+
+/** The next token to be consumed */
+export const NEXT = Symbol("next");
+
 /**
- * Given a state, returns all root states that are reachable from it. We use this
- * to find out which top-level rules want a given token type.
+ * Given a state, returns all root states that are reachable from it, including
+ * partially filled out arguments to the given rule. We use this to find out
+ * which top-level rules want a given token type.
  *
  * @param state The state to get the root states of
+ * @param lastSymbol The partially filled out symbol that is currently being
+ * constructed
+ * @param visitCounts A map of states to the number of times they've been
+ * visited. We use this to prevent infinite recursion
+ * @param roots The list of root states
  * @returns A list of root states
  */
-function getRootStates(state: nearley.State) {
-  /** A queue of states to process; ensures we don't try to process state twice */
-  const queue = new UniqueWorkQueue(state);
-  const roots: State[] = [];
+function computeRootStatePartialArgs(
+  state: nearley.State,
+  lastSymbol: any = NEXT,
+  visitCounts = new DefaultMap<State, number>(() => 0),
+  roots: { state: State; partialArg: any }[] = [],
+) {
+  const visitCount = visitCounts.get(state);
+  if (visitCount > MAX_VISITS) {
```

Comment:
if perf matters, consider wrapping this around the actual call (with condition inverted), to avoid a bunch of leaf recursion calls

Reply:
I think that would only help for things that hit the limit, and be pretty marginal gain, so I'd be inclined to hold off until we decide it matters. Tho I may be missing something here